src/engine/client/cl_main.cpp

```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
+class DemoStopRecordCmd: public Cmd::StaticCmd
 {
-	char name[ MAX_OSPATH ];
-	const char *s;
-
-	if ( Cmd_Argc() > 2 )
-	{
-		Cmd_PrintUsage( "<name>", nullptr );
-		return;
-	}
+public:
+    DemoStopRecordCmd()
+        : Cmd::StaticCmd("demo_record_stop", Cmd::SYSTEM, "Stops recording a demo")
+    {}
 
-	if ( clc.demorecording )
-	{
-		Log::Warn("Already recording." );
-		return;
-	}
+    void Run(const Cmd::Args&) const OVERRIDE
+    {
+        if ( !clc.demorecording )
+        {
+            Log::Notice("%s", "Not recording a demo.\n" );
+            return;
+        }
+        CL_StopRecord();
+    }
+};
+static DemoStopRecordCmd DemoStopRecordCmdRegistration;
 
-	if ( cls.state != connstate_t::CA_ACTIVE )
-	{
-		Log::Warn("You must be in a level to record." );
-		return;
-	}
 
-	// ATVI Wolfenstein Misc #479 - changing this to a warning
-	// sync 0 doesn't prevent recording, so not forcing it off .. everyone does g_sync 1 ; record ; g_sync 0 ..
-	if ( NET_IsLocalAddress( clc.serverAddress ) && !Cvar_VariableValue( "g_synchronousClients" ) )
-	{
-		Log::Warn("You should set '%s' for smoother demo recording" , "g_synchronousClients 1" );
-	}
+class DemoRecordCmd : public Cmd::StaticCmd
+{
+public:
+    DemoRecordCmd()
+        : Cmd::StaticCmd("demo_record", Cmd::SYSTEM, "Begins recording a demo from the current position")
+    {}
 
-	if ( Cmd_Argc() == 2 )
-	{
-		s = Cmd_Argv( 1 );
-		Q_strncpyz( demoName, s, sizeof( demoName ) );
-		Com_sprintf( name, sizeof( name ), "demos/%s.dm_%d", demoName, PROTOCOL_VERSION );
-	}
+    void Run(const Cmd::Args& args) const OVERRIDE
+    {
+        if ( args.size() > 2 )
+        {
+            PrintUsage(args, "[demoname]", "Begins recording a demo from the current position");
+            return;
+        }
 
-	else
-	{
-		char    mapname[ MAX_QPATH ];
-		char    *period;
-		qtime_t time;
+        if ( clc.demorecording )
+        {
+            Log::Warn("Already recording." );
+            return;
+        }
 
-		Com_RealTime( &time );
+        if ( cls.state != connstate_t::CA_ACTIVE )
+        {
+            Log::Warn("You must be in a level to record." );
+            return;
+        }
 
-		Q_strncpyz( mapname, cl.mapname, MAX_QPATH );
+        // ATVI Wolfenstein Misc #479 - changing this to a warning
+        // sync 0 doesn't prevent recording, so not forcing it off .. everyone does g_sync 1 ; record ; g_sync 0 ..
+        if ( NET_IsLocalAddress(clc.serverAddress) && !Cvar_VariableValue("g_synchronousClients") )
+        {
+            Log::Warn("You should set '%s' for smoother demo recording" , "g_synchronousClients 1");
+        }
 
-		for ( period = mapname; *period; period++ )
-		{
-			if ( *period == '.' )
-			{
-				*period = '\0';
-				break;
-			}
-		}
+        std::string demo_name;
 
-		for ( period = mapname; *period; period++ )
-		{
-			if ( *period == '/' )
-			{
-				break;
-			}
-		}
+        if ( args.size() == 2 )
+        {
+            demo_name = args[1];
+        }
```

Comment:
CL_Record is called elsewhere with an automatically-generated name, it used to have the generation code duplicated so I moved it in CL_Record.